include/cetl/pf17/cetlpf.hpp

```diff
@@ -0,0 +1,168 @@
+/// @warning
+/// polyfill headers cannot be used if CETL_H_ERASE is defined.
+///
+/// CETL polyfill headers will provide the CETL type as an aliased standard type in the std namespace when compiling
+/// using older C++ libraries and will automatically provide types from the standard library when compiling
+/// using newer C++ libraries. This allows for causal use of CETL as a true polyfill library but does violate certain
+/// coding standards. As such, for more critical software we recommend not using these headers but including the
+/// types you use from `cetl::pf17` directly.
+///
+/// For example, (TODO: polyfill 17 examples.)
+///
+
+#ifndef CETL_PF17_H_INCLUDED
+#define CETL_PF17_H_INCLUDED
+
+#include "cetl/cetl.hpp"
+#include "cetl/pf17/sys/memory_resource.hpp"
+
+/// @namespace cetl::pf17 This namespace contains C++17 polyfill (pf) types.
+///     The types within this namespace adhere to the C++17 specification and should be drop-in replaceable
+///     with said standard types where they are available. CETL polyfill types may implement a sub-set of
+///     the required functionality but they will not implement non-compliant functionality.
+///
+/// @namespace cetl::pf17::pmr          CETL polyfill types for the standard Polymorphic Memory Resource (PMR)
+///                                     namespace.
+/// @namespace cetl::pf17::pmr::deviant Types or methods that deviate from the C++17 specification.
+///
+#if (__cplusplus >= CETL_CPP_STANDARD_17 && !defined(CETL_DOXYGEN))
+#    include <memory_resource>
+#    include <cstddef>
+
+namespace cetl
+{
+
+using byte = std::byte;
+
+namespace pmr
+{
+
+using memory_resource = std::pmr::memory_resource;
+
+inline memory_resource* null_memory_resource() noexcept
+{
+    return std::pmr::null_memory_resource();
+}
+
+template <typename T>
+using polymorphic_allocator = std::pmr::polymorphic_allocator<T>;
+
+inline memory_resource* new_delete_resource() noexcept
+{
+    return std::pmr::new_delete_resource();
+}
+
+}  // namespace pmr
+}  // namespace cetl
+
+#else
+#    include "cetl/pf17/byte.hpp"
+
+namespace cetl
+{
+
+using byte = cetl::pf17::byte;
+
+namespace pmr
+{
+using memory_resource = cetl::pf17::pmr::memory_resource;
+
+inline memory_resource* null_memory_resource() noexcept
+{
+    return cetl::pf17::pmr::null_memory_resource();
+}
+
+template <typename T>
+using polymorphic_allocator = cetl::pf17::pmr::polymorphic_allocator<T>;
+
+inline memory_resource* new_delete_resource() noexcept
+{
+    return cetl::pf17::pmr::new_delete_resource();
+}
+
+}  // namespace pmr
+}  // namespace cetl
+#endif
+
+#ifdef CETL_PMR_ARRAY_MEMORY_RESOURCE_H_INCLUDED
+
+namespace cetl
+{
+namespace pmr
+{
+
+/// Automatic implementation of cetl::pmr::memory_resource that uses
+/// cetl::pmr::UnsynchronizedArrayMemoryResourceDelegate as the internal implementation.
+/// @tparam UpstreamResourceType The type of any upstream memory resource provided.
+template <typename UpstreamResourceType>
```

Comment:
It's a little weird that this requires a template parameter UpstreamResourceType.  Is this to keep it header-only?  And also I guess UnsynchronizedArrayMemoryResourceDelegate needs a way to refer to the type of the upstream but it can't since memory_resource is ultimately defined here in the polyfill header?

Reply:
That's right. The polyfill type and the inner, delegate support using std::pmr types or cetl::pmr types so this template parameter allows either to be generated from the same logic.